source/D2Common/src/D2Text.cpp

```diff
@@ -175,151 +175,116 @@ short __stdcall TEXT_GetStringIdOfMenuType0Or2Node(D2TextHeaderStrc* pTextHeader
+	D2_ASSERT(pTextHeader != nullptr);
+	D2_ASSERT(pMsgList != nullptr);
 
 	memset(pMsgList, 0x00, sizeof(D2MessageListStrc));
 
-	pNode = pTextHeader->pNode;
 	pMsgList->nCount = LOBYTE(pTextHeader->nCount);
+	
+	int i = 0;
+	for (D2TextNodeStrc* pNode = pTextHeader->pNode; pNode != nullptr; pNode = pNode->pNext) {
+		D2_ASSERT(i < MAX_TEXT_LIST_NODES);
 
-	if (pNode)
-	{
-		do
-		{
-			D2_ASSERT(nCounter < MAX_TEXT_LIST_NODES);
-
-			pMsgList->pMessages[nCounter].nMenu = LOBYTE(pNode->nMenu);
-			pMsgList->pMessages[nCounter].nStringId = pNode->nStringId;
-			++nCounter;
-
-			pNode = pNode->pNext;
-		}
-		while (pNode);
+		pMsgList->pMessages[i].nMenu = LOBYTE(pNode->nMenu);
+		pMsgList->pMessages[i].nStringId = pNode->nStringId;
+		++i;
 	}
 }
 
 //D2Common.0x6FDC3970 (#10910)
 void __stdcall TEXT_CreateTextHeaderFromMessageList(D2TextHeaderStrc* pTextHeader, D2MessageListStrc* pMsgList)
 {
-	D2TextNodeStrc* pNext = NULL;
-	D2TextNodeStrc* pNode = NULL;
+	D2_ASSERT(pTextHeader != nullptr);
+	D2_ASSERT(pMsgList != nullptr);
 
-	D2_ASSERT(pTextHeader);
-	D2_ASSERT(pMsgList);
-
-	pNode = pTextHeader->pNode;
-	if (pNode)
+	D2TextNodeStrc* pNext;
+	for (D2TextNodeStrc* pCurrent = pTextHeader->pNode; pCurrent != nullptr; pCurrent = pNext)
 	{
-		do
-		{
-			pNext = pNode->pNext;
-			FOG_FreeServerMemory(pTextHeader->pMemPool, pNode, __FILE__, __LINE__, 0);
-			pNode = pNext;
-		}
-		while (pNode);
+		pNext = pCurrent->pNext;
+		FOG_FreeServerMemory(pTextHeader->pMemPool, pCurrent, __FILE__, __LINE__, 0);
 	}
 
 	D2_ASSERT(pMsgList->nCount < MAX_TEXT_LIST_NODES);
 
-	pTextHeader->nCount = 0;
-	while (pTextHeader->nCount < pMsgList->nCount)
+	for (pTextHeader->nCount = 0; pTextHeader->nCount < pMsgList->nCount; ++pTextHeader->nCount)
 	{
-		pNode = (D2TextNodeStrc*)FOG_AllocServerMemory(pTextHeader->pMemPool, sizeof(D2TextNodeStrc), __FILE__, __LINE__, 0);
+		D2TextNodeStrc* pNode = (D2TextNodeStrc*)FOG_AllocServerMemory(pTextHeader->pMemPool, sizeof(D2TextNodeStrc), __FILE__, __LINE__, 0);
 		pNode->nMenu = pMsgList->pMessages[pTextHeader->nCount].nMenu;
 		pNode->nStringId = pMsgList->pMessages[pTextHeader->nCount].nStringId;
 		pNode->pNext = pTextHeader->pNode;
-
-		++pTextHeader->nCount;
 		pTextHeader->pNode = pNode;
 	}
 }
 
 //D2Common.0x6FDC3A70 (#10911)
 void __stdcall TEXT_SortTextNodeListByStringId(D2TextHeaderStrc* pTextHeader)
 {
-	D2TextNodeStrc* pNodeList = NULL;
-	D2TextNodeStrc* pNode = NULL;
-	int nPreviousMenu = 0;
-	int nMenu = 0;
-	int nCounter = 0;
-	int nId = 0;
-	short nPreviousStringId = 0;
-	short nStringId = 0;
+	D2_ASSERT(pTextHeader != nullptr);
+
+	// The list is already sorted if it is empty, or has one element.
+	if (pTextHeader->nCount <= 1) {
+		return;
+	}
 
-	D2_ASSERT(pTextHeader);
+	// Copy all of the input into a separate node list for sorting.
+	D2TextNodeStrc* pNodeList = (D2TextNodeStrc*)FOG_AllocServerMemory(pTextHeader->pMemPool, sizeof(D2TextNodeStrc) * pTextHeader->nCount, __FILE__, __LINE__, 0);
```

Comment:
I'm starting to think we should have array versions of `D2_ALLOC_STRC_SERVER`.
Maybe `D2_ALLOC_STRC_N_SERVER(pMemPool, type, count)` or `D2_ALLOC_ARRAY_SERVER(pMemPool, type, count)`.

Your thoughts ? (don't need to add those, I can do it later)